data_mgr.py

```diff
@@ -0,0 +1,296 @@
+#!/usr/bin/env python3
+
+# -*- coding: utf-8 -*-
+# Copyright (C) 2019 NIWA & British Crown (Met Office) & Contributors.
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+"""Create and update the data structure for all workflow services."""
+
+from fnmatch import fnmatchcase
+
+from tornado import gen
+# from google.protobuf.json_format import MessageToDict
+
+# from cylc import flags
+from cylc.exceptions import ClientError, ClientTimeout
+from cylc.ws_messages_pb2 import PbEntireWorkflow
+
+MSG_QUIT = "QUIT"
+MSG_TIMEOUT = "TIMEOUT"
+CYCLING_TYPES = [
+    'task_proxies',
+    'family_proxies',
+    'jobs',
+]
+METHOD_MAP = {
+    'pb_entire_workflow': PbEntireWorkflow
+}
+
+
+async def get_workflow_data(id, client, method):
+    # Use already established client
+    try:
+        pb_msg = await client.async_request(method)
+    except ClientTimeout as exc:
+        return (id, MSG_TIMEOUT)
+    except ClientError as exc:
+        return (id, None)
+    else:
+        ws_data = METHOD_MAP[method]()
+        ws_data.ParseFromString(pb_msg)
+    return (id, ws_data)
+
+
+class DataManager(object):
+
+    def __init__(self, ws_mgr):
+        self.ws_mgr = ws_mgr
+        self.data = {}
+
+    # **
+    # Data syncing
+    # **
+    async def entire_workflow_update(self, ids=[]):
```

Comment:
Mutable argument `ids`. Instead it should be `None` and then check its value... see [this](https://docs.python-guide.org/writing/gotchas/#mutable-default-arguments) one for more.

Reply:
Surprising behaviour.. Thanks!